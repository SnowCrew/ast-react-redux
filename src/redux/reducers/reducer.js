

const initialState = {
  products: [],
  loading: true,
  error: false,
  itemsInCart: [],
  total: 0
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'MENU_LOADED':
      return {
        ...state,
        products: action.payload,
        loading: false,
        error: false
      };
    case 'MENU_REQUESTED':
      return {
        ...state,
        products: state.products,
        loading: true,
        error: false
      };
    case 'MENU_ERROR':
      return {
        ...state,
        products: state.products,
        loading: true,
        error: true
      };
    case "ITEM_ADD_TO_CART":
      const id = action.payload;

      const itemInd = state.itemsInCart.findIndex(item => item.id === id);

      if (itemInd >= 0) {
        const elemInState = state.itemsInCart.find(item => item.id === id);
        const newElem = {
          ...elemInState,
          counterIdentical: ++elemInState.counterIdentical
        }
        return {
          ...state,
          itemsInCart: [
            ...state.itemsInCart.slice(0, itemInd),
            newElem,
            ...state.itemsInCart.slice(itemInd + 1)
          ],
          total: (Number(state.total) + Number(newElem.price)).toFixed(2)
        }
      }

      const item = state.products.find(item => item.id === id);
      const newItem = {
        title: item.title,
        price: item.price,
        image: item.image,
        id: item.id,
        counterIdentical: 1,
      }
      return {
        ...state,
        itemsInCart: [
          ...state.itemsInCart,
          newItem
        ],
        total: (Number(state.total) + Number(newItem.price)).toFixed(2)
      };

    case "ITEM_REMOVE_FROM_CART":
      const idx = action.payload;
      const itemIndex = state.itemsInCart.findIndex(item => item.id === idx);
      const itemDel = state.itemsInCart.find(item => item.id === idx);
      return {
        ...state,
        itemsInCart: [
          ...state.itemsInCart.slice(0, itemIndex),
          ...state.itemsInCart.slice(itemIndex + 1)
        ],
        total: state.total - (itemDel.price * itemDel.counterIdentical)
      };

    default:
      return state;
  }
}

export default reducer;