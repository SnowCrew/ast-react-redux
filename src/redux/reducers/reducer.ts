import { IAction } from '../actions/actions';


const initialState:IStore = {
  products: [],
  loading: true,
  error: false,
  itemsInCart: [],
  total: 0,
  totalQuantity: 0,
  auth: false,
  admin: false
};

export interface IProduct {
  id: number,
  title:string,
  price:number,
  category:string,
  description:string,
  image:string
}

export interface IProductInCart extends IProduct {
counterIdentical:number,
}

export interface IStore {
  products: IProduct[],
  loading: boolean,
  error: boolean,
  itemsInCart: IProductInCart[],
  total: number,
  totalQuantity: number,
  auth: boolean,
  admin: boolean
};

const reducer = (state:IStore = initialState, action:IAction):IStore => {
  switch (action.type) {
    case 'MENU_LOADED':
      return {
        ...state,
        products: action.payload as IProduct[],
        loading: false,
        error: false
      };
    case 'MENU_REQUESTED':
      return {
        ...state,
        products: state.products,
        loading: true,
        error: false
      };
    case 'MENU_ERROR':
      return {
        ...state,
        products: state.products,
        loading: true,
        error: true
      };
    case "ITEM_ADD_TO_CART":
      const [id, quantity] = action.payload as [number, number];
      const itemInd = state.itemsInCart?.findIndex(item => item.id === id);

      if (itemInd >= 0 ) {
        const elemInState = state.itemsInCart?.find(item => item.id === id);
        const newElem = {
          ...elemInState,
          counterIdentical: (Number(elemInState?.counterIdentical) + Number(quantity)),
        }
        return {
          ...state,
          itemsInCart: [
            ...state.itemsInCart!.slice(0, itemInd),
            newElem,
            ...state.itemsInCart!.slice(itemInd! + 1)
          ] as IProductInCart[],
          total: Number((Number(state.total) + (Number(newElem.price) * Number(quantity))).toFixed(2)),
          totalQuantity: Number(state.totalQuantity) + Number(quantity)
        }
      }

      const item = state.products.find(item => item.id === id);
      const newItem:IProductInCart = {
        title: item!.title,
        price: item!.price,
        image: item!.image,
        id: item!.id,
        category:item!.category,
        description: item!.description,
        counterIdentical: Number(quantity)
      }
      return {
        ...state,
        itemsInCart: [
          ...state.itemsInCart,
          newItem
        ],
        total: Number(((state.total) + (Number(newItem.price) * Number(quantity))).toFixed(2)),
        totalQuantity: Number(state.totalQuantity) + Number(quantity)
      };

    case "ITEM_REMOVE_FROM_CART":
      const idx = action.payload as number;

      const itemIndex = state.itemsInCart.findIndex(item => item.id === idx);
      const itemDel = state.itemsInCart.find(item => item.id === idx);
      if(itemIndex === -1){
        return state;
      }
      return {
        ...state,
        itemsInCart: [
          ...state.itemsInCart.slice(0, itemIndex),
          ...state.itemsInCart.slice(itemIndex + 1)
        ],
        total: Number((state.total - (itemDel!.price * itemDel!.counterIdentical)).toFixed(2)),
        totalQuantity: Number(state.totalQuantity) - itemDel!.counterIdentical
      };

    case "CLEAR_CART":
      return {
        ...state,
        itemsInCart: [],
        total: 0,
        totalQuantity: 0
      }

    case "GET_AUTH_REQUEST":
      return {
        ...state,
        loading: true
      };

    case "GET_AUTH":
      return {
        ...state,
        auth: true,
        loading: false
      };

    case "GET_ADMIN_AUTH":
      return {
        ...state,
        auth: true,
        loading: false,
        admin: true
      };

    case "GET_AUTH_REJECT":
      return {
        ...state,
        loading: false
      };

    case "EXIT_AUTH":
      return {
        ...state,
        auth: false,
        admin: false
      };

    default:
      return state;
  };
};

export default reducer;