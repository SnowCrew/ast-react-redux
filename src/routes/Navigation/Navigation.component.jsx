import "./Navigation.styles.css";
import { Link, Outlet } from "react-router-dom";
import CartIcon from "./shopping-cart-solid.svg";
import AuthComponent from "../../components/Auth/Auth.component";

const Navigation = () => {
  return (
    <>
      <div className="header">
        <div className="navigation">
          <Link className="nav-link" to="/">
            Главная
          </Link>
          <Link className="nav-link" to="/about">
            О магазине
          </Link>
        </div>
        <AuthComponent CartIcon={CartIcon} />
      </div>
      <Outlet />
    </>
  );
};

export default Navigation;
