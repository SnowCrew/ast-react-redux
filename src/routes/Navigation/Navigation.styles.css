.header {
  position: sticky;
  top: 0;
  z-index: 1;

  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;

  margin-bottom: 25px;

  height: 50px;

  background-color: rgb(191, 246, 246);
}
.nav-link {
  margin: 0 25px;
  text-decoration: none;
  color: black;
}

.nav-cart {
  width: 20px;
  margin-right: 10px;
  /* fill: rgb(190, 20, 20); */
}
