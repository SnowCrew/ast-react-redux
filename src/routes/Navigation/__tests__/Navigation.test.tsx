import { cleanup, render, screen } from "@testing-library/react";
import { ReactElement } from "react";
import { legacy_createStore as createStore } from "redux";
import reducer, { IProduct } from "../../../redux/reducers/reducer";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import "@testing-library/jest-dom/extend-expect";
import Navigation from "../Navigation.component";

const renderWithRedux = (
  component: ReactElement,
  { initialState, store = createStore(reducer, initialState) } = {}
) => {
  return {
    ...render(<Provider store={store}>{component}</Provider>, {
      wrapper: MemoryRouter,
    }),
  };
};

describe("Navigation render without errors", () => {
  it("Navigation rendering", () => {
    expect(renderWithRedux(<Navigation />)).toMatchSnapshot();
  });
});
