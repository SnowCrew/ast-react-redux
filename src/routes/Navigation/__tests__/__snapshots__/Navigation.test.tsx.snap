// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartPage render without errors CartPage rendering 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="header"
      >
        <div
          class="navigation"
        >
          <a
            class="nav-link"
            href="/"
          >
            Главная
          </a>
          <a
            class="nav-link"
            href="/about"
          >
            О магазине
          </a>
        </div>
        <div
          class="auth-block"
        >
          <span
            class="enter-exit-span"
            data-testid="set_modal_active_button"
          >
            Вход
          </span>
          <div
            class="modal"
          >
            <div
              class="modal__content"
            >
              <div
                class="auth-modal-block"
              >
                <h2
                  class="auth-title"
                >
                  Авторизация
                </h2>
                <label
                  class="auth-label"
                >
                  Email
                </label>
                <input
                  class="auth-input"
                  placeholder="Enter email"
                  type="email"
                />
                <label
                  class="auth-label"
                >
                  Password
                </label>
                <input
                  class="auth-input"
                  placeholder="Enter password"
                  type="text"
                />
                <button
                  class="auth-btn"
                >
                  Войти
                </button>
                <button
                  class="auth-btn"
                >
                   
                  Заполнить поля коррект. данными юзера
                </button>
                <button
                  class="auth-btn"
                >
                  Заполнить поля данными админа
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="header"
    >
      <div
        class="navigation"
      >
        <a
          class="nav-link"
          href="/"
        >
          Главная
        </a>
        <a
          class="nav-link"
          href="/about"
        >
          О магазине
        </a>
      </div>
      <div
        class="auth-block"
      >
        <span
          class="enter-exit-span"
          data-testid="set_modal_active_button"
        >
          Вход
        </span>
        <div
          class="modal"
        >
          <div
            class="modal__content"
          >
            <div
              class="auth-modal-block"
            >
              <h2
                class="auth-title"
              >
                Авторизация
              </h2>
              <label
                class="auth-label"
              >
                Email
              </label>
              <input
                class="auth-input"
                placeholder="Enter email"
                type="email"
              />
              <label
                class="auth-label"
              >
                Password
              </label>
              <input
                class="auth-input"
                placeholder="Enter password"
                type="text"
              />
              <button
                class="auth-btn"
              >
                Войти
              </button>
              <button
                class="auth-btn"
              >
                 
                Заполнить поля коррект. данными юзера
              </button>
              <button
                class="auth-btn"
              >
                Заполнить поля данными админа
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Navigation render without errors Navigation rendering 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="header"
      >
        <div
          class="navigation"
        >
          <a
            class="nav-link"
            href="/"
          >
            Главная
          </a>
          <a
            class="nav-link"
            href="/about"
          >
            О магазине
          </a>
        </div>
        <div
          class="auth-block"
        >
          <span
            class="enter-exit-span"
            data-testid="set_modal_active_button"
          >
            Вход
          </span>
          <div
            class="modal"
          >
            <div
              class="modal__content"
            >
              <div
                class="auth-modal-block"
              >
                <h2
                  class="auth-title"
                >
                  Авторизация
                </h2>
                <label
                  class="auth-label"
                >
                  Email
                </label>
                <input
                  class="auth-input"
                  placeholder="Enter email"
                  type="email"
                />
                <label
                  class="auth-label"
                >
                  Password
                </label>
                <input
                  class="auth-input"
                  placeholder="Enter password"
                  type="text"
                />
                <button
                  class="auth-btn"
                >
                  Войти
                </button>
                <button
                  class="auth-btn"
                >
                   
                  Заполнить поля коррект. данными юзера
                </button>
                <button
                  class="auth-btn"
                >
                  Заполнить поля данными админа
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="header"
    >
      <div
        class="navigation"
      >
        <a
          class="nav-link"
          href="/"
        >
          Главная
        </a>
        <a
          class="nav-link"
          href="/about"
        >
          О магазине
        </a>
      </div>
      <div
        class="auth-block"
      >
        <span
          class="enter-exit-span"
          data-testid="set_modal_active_button"
        >
          Вход
        </span>
        <div
          class="modal"
        >
          <div
            class="modal__content"
          >
            <div
              class="auth-modal-block"
            >
              <h2
                class="auth-title"
              >
                Авторизация
              </h2>
              <label
                class="auth-label"
              >
                Email
              </label>
              <input
                class="auth-input"
                placeholder="Enter email"
                type="email"
              />
              <label
                class="auth-label"
              >
                Password
              </label>
              <input
                class="auth-input"
                placeholder="Enter password"
                type="text"
              />
              <button
                class="auth-btn"
              >
                Войти
              </button>
              <button
                class="auth-btn"
              >
                 
                Заполнить поля коррект. данными юзера
              </button>
              <button
                class="auth-btn"
              >
                Заполнить поля данными админа
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
